compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -723,6 +723,138 @@ OMR::ARM64::TreeEvaluator::lremEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    return iremHelper(node, true, cg);
    }
 
+/**
+ * @brief Generates ubfm instruction for mask and shift operation if possible
+ *
+ * @param[in] shiftNode: node with shift opcode
+ * @param[in]        cg: code generator
+ *
+ * @returns register which contains the result of the operation. NULL if the operation cannot be encoded in ubfm instruction.
+ */
+static TR::Register *
+generateUBFMForMaskAndShift(TR::Node *shiftNode, TR::CodeGenerator *cg)
+   {
+   TR::Node *andNode = NULL;
+   TR::Node *shiftValueNode = NULL;
+   TR::Node *maskNode = NULL;
+   TR::Node *sourceNode = NULL;
+   if (shiftNode->getFirstChild()->getOpCode().isAnd())
+      {
+      andNode = shiftNode->getFirstChild();
+      shiftValueNode = shiftNode->getSecondChild();
+      if (!shiftValueNode->getOpCode().isLoadConst())
+         {
+         return NULL;
+         }
+      if ((andNode->getReferenceCount() > 1) || (andNode->getRegister() != NULL))
+         {
+         return NULL;
+         }
+      if (andNode->getSecondChild()->getOpCode().isLoadConst())
+         {
+         sourceNode = andNode->getFirstChild();
+         maskNode = andNode->getSecondChild();
+         }
+      else if (andNode->getFirstChild()->getOpCode().isLoadConst())
+         {
+         sourceNode = andNode->getSecondChild();
+         maskNode = andNode->getFirstChild();
+         }
+      }
+
+   if (maskNode == NULL)
+      {
+      return NULL;
+      }
+   const bool is64bit = shiftNode->getDataType().isInt64();
+   const int64_t shiftValue = shiftValueNode->getConstValue();
+
+   if ((shiftValue <= 0) || (shiftValue > (is64bit ? 63 : 31)))
+      {
+      return NULL;
+      }
+   const uint64_t maskValue = is64bit ? maskNode->getLongInt() : static_cast<uint32_t>(maskNode->getInt());
+
+   if (shiftNode->getOpCode().isLeftShift())
+      {
+      if ((maskValue << shiftValue) == 0)
+         {
+         /*
+          * The result is always 0 in this case.
+          */
+         TR::Register *reg = cg->allocateRegister();
+         generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, shiftNode, reg, 0);
+         shiftNode->setRegister(reg);
+         cg->recursivelyDecReferenceCount(andNode);
+         cg->decReferenceCount(shiftValueNode);
+         return reg;
+         }
+
+      if (((maskValue & 1) == 1) && (((maskValue >> (is64bit ? 63 : 31)) & 1) == 0) && contiguousBits(maskValue))
+         {
+         /*
+          * maskValue has consecutive 1s from the least significant bits. The most significant bit must be 0. Otherwise, mask is all 1s.
+          * So, this mask and shift operation is copying consecutive bits starting from the least significant bits of the source register
+          * to bit position shiftValue of the destination register.
+          */
+         uint32_t width = populationCount(maskValue);
+         uint32_t imms = (width - 1);
+         uint32_t immr = (is64bit ? 64 : 32) - shiftValue;
+
+         TR::Register *reg = cg->allocateRegister();
```

Comment:
Please consider checking the reference count of `sourceNode`, to reuse `sreg` as `treg`.

Reply:
force-pushed the updated commit to address konno-san's comment.